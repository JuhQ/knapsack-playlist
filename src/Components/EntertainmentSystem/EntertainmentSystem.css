h2 {
  background-image: linear-gradient(
    69deg,
    var(--color-three),
    var(--color-four),
    var(--color-two),
    var(--color-one)
  );
  background-clip: text;
  -webkit-text-fill-color: transparent;
  transform: all 0.3s ease;
  margin: 2rem;
}

.done {
  font-size: 3rem;
}
