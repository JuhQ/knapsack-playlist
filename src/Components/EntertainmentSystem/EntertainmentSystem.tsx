import "./EntertainmentSystem.css";

import React, { useEffect, useState } from "react";

import Queue from "../../Datastructures/queue/queue";
import { YoutubeItem } from "../../types";
import EntertainmentSystemContent from "./EntertainmentSystemContent/EntertainmentSystemContent";
import EntertainmentSystemLastSongPlayed from "./EntertainmentSystemLastSongPlayed/EntertainmentSystemLastSongPlayed";

interface Props {
  list: Queue;
  length: number;
}

const EntertainmentSystem: React.FC<Props> = ({ list, length }: Props) => {
  const [playlist, setPlaylist] = useState<Queue>(list);
  const [currentSong, setCurrentSong] = useState<YoutubeItem | undefined>(
    list.at(0)
  );
  const [lastSongPlayed, setLastSongPlayed] = useState<boolean>(false);
  const playlistLength = list.seconds();
  const diff = length - playlistLength;

  useEffect(() => {
    const initialPlaylist = new Queue();
    initialPlaylist.merge(list);
    const firstSong = initialPlaylist.dequeue();

    setCurrentSong(firstSong);

    setPlaylist(initialPlaylist);
    setLastSongPlayed(false);
  }, [list]);

  if (lastSongPlayed) {
    return <EntertainmentSystemLastSongPlayed />;
  }

  return (
    <EntertainmentSystemContent
      list={list}
      playlist={playlist}
      currentSong={
        currentSong || { id: "1", title: "1", seconds: 0, rating: 1 }
      }
      length={length}
      diff={diff}
      playlistLength={playlistLength}
      onEnd={() => {
        const newQueue = new Queue();
        newQueue.merge(playlist);

        const nextSong = newQueue.dequeue();

        if (nextSong) {
          setCurrentSong(nextSong);
          setPlaylist(newQueue);
        } else {
          setLastSongPlayed(true);
          setPlaylist(new Queue());
        }
      }}
      onChange={(song) => {
        const queue = new Queue();
        const filteredQueue = playlist.all().filter(({ id }) => id !== song.id);
        queue.merge(filteredQueue);
        setPlaylist(queue);
        setCurrentSong(song);
      }}
    />
  );
};

export default EntertainmentSystem;
