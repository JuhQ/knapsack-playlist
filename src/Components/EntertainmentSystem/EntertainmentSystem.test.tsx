import { shallow } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import React from "react";
import { act, create } from "react-test-renderer";

import Queue from "../../Datastructures/queue/queue";
import EntertainmentSystem from "./EntertainmentSystem";

describe("EntertainmentSystem", () => {
  it("should render", () => {
    const list = new Queue();
    list.merge([
      { id: "1", title: "test", seconds: 60 },
      { id: "2", title: "test 2", seconds: 60 },
    ]);
    const component = create(<EntertainmentSystem list={list} length={120} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    act(() => {
      expect(tree).toMatchSnapshot();
    });
  });

  it("should end the first song", () => {
    const list = new Queue();
    list.merge([
      { id: "1", title: "test", seconds: 60 },
      { id: "2", title: "test 2", seconds: 60 },
    ]);
    const wrapper = shallow(<EntertainmentSystem list={list} length={120} />);

    wrapper.find("EntertainmentSystemContent").simulate("end");
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should end the last song of the list", () => {
    const list = new Queue();
    list.merge([{ id: "1", title: "test", seconds: 60 }]);

    const wrapper = shallow(<EntertainmentSystem list={list} length={120} />);
    wrapper.find("EntertainmentSystemContent").simulate("end");
    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });

  it("should render empty playlist message", () => {
    const wrapper = shallow(
      <EntertainmentSystem list={new Queue()} length={120} />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it("should end song on empty list", () => {
    const wrapper = shallow(
      <EntertainmentSystem list={new Queue()} length={120} />
    );

    wrapper.find("EntertainmentSystemContent").simulate("end");
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it("should change song", () => {
    const list = new Queue();
    list.merge([
      { id: "1", title: "test", seconds: 60 },
      { id: "2", title: "test 2", seconds: 60 },
    ]);
    const wrapper = shallow(<EntertainmentSystem list={list} length={120} />);

    wrapper
      .find("EntertainmentSystemContent")
      .simulate("change", { id: "2", title: "test 2", seconds: 60 });
    act(() => {
      expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
  });
});
