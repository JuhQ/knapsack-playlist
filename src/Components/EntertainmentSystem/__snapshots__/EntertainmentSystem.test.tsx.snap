// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntertainmentSystem should change song 1`] = `
<EntertainmentSystemContent
  currentSong={
    Object {
      "id": "2",
      "rating": 1,
      "seconds": 60,
      "title": "test 2",
    }
  }
  diff={0}
  length={120}
  list={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 2,
        "list": Array [
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          },
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          },
        ],
      },
    }
  }
  onChange={[Function]}
  onEnd={[Function]}
  playlist={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 1,
        "list": Array [
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          },
        ],
      },
    }
  }
  playlistLength={120}
/>
`;

exports[`EntertainmentSystem should end song on empty list 1`] = `<EntertainmentSystemLastSongPlayed />`;

exports[`EntertainmentSystem should end the first song 1`] = `
<EntertainmentSystemContent
  currentSong={
    Object {
      "id": "1",
      "rating": 1,
      "seconds": 60,
      "title": "test",
    }
  }
  diff={0}
  length={120}
  list={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 2,
        "list": Array [
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          },
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          },
        ],
      },
    }
  }
  onChange={[Function]}
  onEnd={[Function]}
  playlist={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 1,
        "list": Array [
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          },
        ],
      },
    }
  }
  playlistLength={120}
/>
`;

exports[`EntertainmentSystem should end the first song 2`] = `
<EntertainmentSystemContent
  currentSong={
    Object {
      "id": "1",
      "rating": 1,
      "seconds": 60,
      "title": "test",
    }
  }
  diff={0}
  length={120}
  list={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 2,
        "list": Array [
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          },
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          },
        ],
      },
    }
  }
  onChange={[Function]}
  onEnd={[Function]}
  playlist={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 1,
        "list": Array [
          Object {
            "id": "2",
            "rating": 1,
            "seconds": 60,
            "title": "test 2",
          },
        ],
      },
    }
  }
  playlistLength={120}
/>
`;

exports[`EntertainmentSystem should end the last song of the list 1`] = `
<EntertainmentSystemContent
  currentSong={
    Object {
      "id": "1",
      "rating": 1,
      "seconds": 60,
      "title": "test",
    }
  }
  diff={60}
  length={120}
  list={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 1,
        "list": Array [
          Object {
            "id": "1",
            "rating": 1,
            "seconds": 60,
            "title": "test",
          },
        ],
      },
    }
  }
  onChange={[Function]}
  onEnd={[Function]}
  playlist={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 0,
        "list": Array [],
      },
    }
  }
  playlistLength={60}
/>
`;

exports[`EntertainmentSystem should render 1`] = `
Array [
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        2
         songs
      </span>
    </div>
  </div>,
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/1/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test 2
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`EntertainmentSystem should render 2`] = `
Array [
  <div
    className="ui bottom aligned grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <h2>
        test
      </h2>
    </div>
    <div
      className="center aligned five wide computer sixteen wide mobile column"
    >
      <span>
        Playlist duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Tried to generate 
        120
         seconds
      </span>
      <br />
      <strong>
        Great success! Playlist is exactly as long as you wanted!
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ‘Œ
        </span>
        <span
          aria-label="Success"
          role="img"
        >
          ðŸ¤˜
        </span>
        <br />
      </strong>
      <span>
        Queue duration: 
        120
         seconds
      </span>
      <br />
      <span>
        Playlist size: 
        2
         songs
      </span>
      <br />
      <span>
        Queue size: 
        2
         songs
      </span>
    </div>
  </div>,
  <div
    className="ui grid"
  >
    <div
      className="right aligned ten wide computer sixteen wide mobile column"
    >
      <div
        className=""
      >
        <div
          className={null}
          id={null}
        />
      </div>
    </div>
    <div
      className="five wide computer sixteen wide mobile column"
    >
      <div
        className="playlist"
      >
        <div
          className="ui items"
        >
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/1/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
          <a
            className="item"
            onClick={[Function]}
          >
            <div
              className="image"
            >
              <img
                src="https://i.ytimg.com/vi/2/hqdefault.jpg"
              />
            </div>
            <div
              className="content"
            >
              <div
                className="header"
              >
                test 2
              </div>
              <div
                className="meta"
              >
                1 minute
                 
                
              </div>
            </div>
          </a>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`EntertainmentSystem should render empty playlist message 1`] = `
<EntertainmentSystemContent
  currentSong={
    Object {
      "id": "1",
      "rating": 1,
      "seconds": 0,
      "title": "1",
    }
  }
  diff={120}
  length={120}
  list={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 0,
        "list": Array [
          undefined,
        ],
      },
    }
  }
  onChange={[Function]}
  onEnd={[Function]}
  playlist={
    Queue {
      "enqueue": [Function],
      "list": ArrayList {
        "index": 0,
        "list": Array [
          undefined,
        ],
      },
    }
  }
  playlistLength={0}
/>
`;
